periodic/src/socket_rollback_handler.rs

```diff
@@ -0,0 +1,334 @@
+pub struct SocketRollbackHandler<T> {
+	/// The `EthClient` to interact with the connected blockchain.
+	pub client: Arc<EthClient<T>>,
+	/// The entire clients instantiated in the system. <chain_id, Arc<EthClient>>
+	system_clients: BTreeMap<ChainID, Arc<EthClient<T>>>,
+	/// The receiver connected to the socket rollback channel.
+	rollback_receiver: UnboundedReceiver<SocketMessage>,
+	/// The local storage saving emitted `Socket` event messages.
+	rollback_msgs: BTreeMap<RawRequestID, RollbackableMessage>,
+	/// The event senders that sends messages to the event channel.
+	event_sender: Arc<EventSender>,
+	/// The time schedule that represents when to check heartbeat pulsed.
+	schedule: Schedule,
+}
+
+impl<T: JsonRpcClient> SocketRollbackHandler<T> {
+	/// Instantiates a new `SocketRollbackHandler`.
+	pub fn new(
+		event_sender: Arc<EventSender>,
+		system_clients_vec: Vec<Arc<EthClient<T>>>,
+	) -> (Self, UnboundedSender<SocketMessage>) {
+		let (sender, rollback_receiver) = mpsc::unbounded_channel::<SocketMessage>();
+
+		let mut system_clients = BTreeMap::new();
+		system_clients_vec.iter().for_each(|client| {
+			system_clients.insert(client.get_chain_id(), client.clone());
+		});
+
+		(
+			Self {
+				client: system_clients.get(&event_sender.id).expect(INVALID_CHAIN_ID).clone(),
+				system_clients,
+				rollback_receiver,
+				rollback_msgs: BTreeMap::new(),
+				event_sender,
+				schedule: Schedule::from_str(ROLLBACK_CHECK_SCHEDULE)
+					.expect(INVALID_PERIODIC_SCHEDULE),
+			},
+			sender,
+		)
+	}
+
+	/// Verifies whether the given socket message has been executed.
+	async fn is_request_executed(&self, socket_msg: &SocketMessage) -> bool {
+		let src_request = self
+			.get_socket_request(&socket_msg.req_id, ChainID::from_be_bytes(socket_msg.req_id.chain))
+			.await;
+		let dst_request = self
+			.get_socket_request(
+				&socket_msg.req_id,
+				ChainID::from_be_bytes(socket_msg.ins_code.chain),
+			)
+			.await;
+
+		if let (Some(src_request), Some(dst_request)) = (src_request, dst_request) {
+			let src_status = SocketEventStatus::from_u8(src_request.field[0].clone().into());
+			let dst_status = SocketEventStatus::from_u8(dst_request.field[0].clone().into());
+
+			match src_status {
+				SocketEventStatus::Committed | SocketEventStatus::Rollbacked => return true,
+				_ => (),
+			}
+			if self.is_inbound_sequence(ChainID::from_be_bytes(socket_msg.ins_code.chain)) {
+				match dst_status {
+					SocketEventStatus::Executed
+					| SocketEventStatus::Reverted
+					| SocketEventStatus::Accepted
+					| SocketEventStatus::Rejected => return true,
+					_ => (),
+				}
+			} else {
+				match dst_status {
+					SocketEventStatus::Executed | SocketEventStatus::Reverted => return true,
+					_ => (),
+				}
+			}
+		}
+		false
+	}
+
+	/// Verifies whether the socket event is an inbound sequence.
+	fn is_inbound_sequence(&self, dst_chain_id: ChainID) -> bool {
+		if let Some(client) = &self.system_clients.get(&dst_chain_id) {
```

Comment:
```suggestion
		if let Some(client) = self.system_clients.get(&dst_chain_id) {
```
변경 전 client의 타입은 `&&Arc<Client<T>>`. 변경 후 `&Arc<Client<T>>`